File: src/main/java/com/danyazero/utils/IntegerType.java
package com.danyazero.utils;

import com.danyazero.model.NumberType;
import org.objectweb.asm.MethodVisitor;

import static org.objectweb.asm.Opcodes.*;

public class IntegerType implements NumberType<Integer> {

    @Override
    public void add(MethodVisitor mv) {
        mv.visitInsn(IADD);
    }

    @Override
    public void sub(MethodVisitor mv) {
        mv.visitInsn(ISUB);
    }

    @Override
    public void mul(MethodVisitor mv) {
        mv.visitInsn(IMUL);
    }

    @Override
    public void div(MethodVisitor mv) {
        mv.visitInsn(IDIV);
    }

    @Override
    public void negate(MethodVisitor mv) {
        mv.visitInsn(INEG);
    }

    @Override
    public int getType() {
        return T_INT;
    }

    @Override
    public void postack(MethodVisitor mv, java.lang.Integer value) {
        if (value >= -1 && value <= 5) {
            switch (value) {
                case -1 -> mv.visitInsn(ICONST_M1);
                case 0 -> mv.visitInsn(ICONST_0);
                case 1 -> mv.visitInsn(ICONST_1);
                case 2 -> mv.visitInsn(ICONST_2);
                case 3 -> mv.visitInsn(ICONST_3);
                case 4 -> mv.visitInsn(ICONST_4);
                case 5 -> mv.visitInsn(ICONST_5);
            }
        } else if (value >= -128 && value <= 127) {
            mv.visitIntInsn(BIPUSH, value);
        } else if ((value >= -32768 && value <= 32767)) {
            mv.visitIntInsn(SIPUSH, value);
        } else {
            mv.visitLdcInsn(value);
        }
    }

    @Override
    public void store(MethodVisitor mv, short index) {
        mv.visitVarInsn(ISTORE, index);
    }

    @Override
    public void load(MethodVisitor mv, short index) {
        mv.visitIntInsn(ILOAD, index);
    }

    @Override
    public void yield(MethodVisitor mv) {
        mv.visitInsn(IRETURN);
    }

    @Override
    public void aload(MethodVisitor mv) {
        mv.visitInsn(IALOAD);
    }

    @Override
    public void astore(MethodVisitor mv) {
        mv.visitInsn(IASTORE);
    }

    @Override
    public short getSize() {
        return 1;
    }
}

File: src/main/kotlin/type/IntegerType.kt
package com.danyazero.type

import com.danyazero.model.NumberType
import org.objectweb.asm.Label
import org.objectweb.asm.MethodVisitor
import org.objectweb.asm.Opcodes

class IntegerType : NumberType<Int> {

    override fun add(mv: MethodVisitor) {
        mv.visitInsn(Opcodes.IADD)
    }

    override fun sub(mv: MethodVisitor) {
        mv.visitInsn(Opcodes.ISUB)
    }

    override fun mul(mv: MethodVisitor) {
        mv.visitInsn(Opcodes.IMUL)
    }

    override fun div(mv: MethodVisitor) {
        mv.visitInsn(Opcodes.IDIV)
    }

    fun inc(mv: MethodVisitor, index: Short, const: Int) {
        mv.visitIincInsn(index.toInt(), const)
    }

    override fun xor(mv: MethodVisitor) {
        mv.visitInsn(Opcodes.IXOR)
    }

    override fun mod(mv: MethodVisitor) {
        mv.visitInsn(Opcodes.IREM)
    }

    override fun ushr(mv: MethodVisitor) {
        mv.visitInsn(Opcodes.ISHR)
    }

    override fun shr(mv: MethodVisitor) {
        mv.visitInsn(Opcodes.ISHR)
    }

    override fun shl(mv: MethodVisitor) {
        mv.visitInsn(Opcodes.ISHL)
    }

    override fun negate(mv: MethodVisitor) {
        mv.visitInsn(Opcodes.INEG)
    }

    override fun postack(mv: MethodVisitor, value: Int) {
        if (value >= -1 && value <= 5) {
            when (value) {
                -1 -> mv.visitInsn(Opcodes.ICONST_M1)
                0 -> mv.visitInsn(Opcodes.ICONST_0)
                1 -> mv.visitInsn(Opcodes.ICONST_1)
                2 -> mv.visitInsn(Opcodes.ICONST_2)
                3 -> mv.visitInsn(Opcodes.ICONST_3)
                4 -> mv.visitInsn(Opcodes.ICONST_4)
                5 -> mv.visitInsn(Opcodes.ICONST_5)
            }
        } else if (value >= -128 && value <= 127) {
            mv.visitIntInsn(Opcodes.BIPUSH, value)
        } else if ((value >= -32768 && value <= 32767)) {
            mv.visitIntInsn(Opcodes.SIPUSH, value)
        } else {
            mv.visitLdcInsn(value)
        }
    }

    override fun store(mv: MethodVisitor, index: Short) {
        mv.visitVarInsn(Opcodes.ISTORE, index.toInt())
    }

    override fun load(mv: MethodVisitor, index: Short) {
        mv.visitIntInsn(Opcodes.ILOAD, index.toInt())
    }

    override fun yield(mv: MethodVisitor) {
        mv.visitInsn(Opcodes.IRETURN)
    }

    override fun aload(mv: MethodVisitor) {
        mv.visitInsn(Opcodes.IALOAD)
    }

    override fun astore(mv: MethodVisitor) {
        mv.visitInsn(Opcodes.IASTORE)
    }

    override fun equal(mv: MethodVisitor, jumpTarget: Label) {
        mv.visitJumpInsn(Opcodes.IF_ICMPNE, jumpTarget)
    }

    override fun greater(mv: MethodVisitor, jumpTarget: Label) {
        mv.visitJumpInsn(Opcodes.IF_ICMPLE, jumpTarget)
    }

    override fun less(mv: MethodVisitor, jumpTarget: Label) {
        mv.visitJumpInsn(Opcodes.IF_ICMPGE, jumpTarget)
    }

    override fun greaterOrEqual(mv: MethodVisitor, jumpTarget: Label) {
        mv.visitJumpInsn(Opcodes.IF_ICMPLT, jumpTarget)
    }

    override fun lessOrEqual(mv: MethodVisitor, jumpTarget: Label) {
        mv.visitJumpInsn(Opcodes.IF_ICMPGT, jumpTarget)
    }

    override fun getSize(): Short {
        return 1
    }

    override fun getType() : Int {
        return Opcodes.T_INT
    }
}